builtin/credential/jwt/path_oidc.go

```diff
@@ -194,28 +235,45 @@ func (b *jwtAuthBackend) pathCallback(ctx context.Context, req *logical.Request,
 
 	stateID := d.Get("state").(string)
 
-	oidcReq := b.verifyOIDCRequest(stateID)
-	if oidcReq == nil {
+	oidcReq := b.getOIDCRequest(stateID)
+	if oidcReq == nil || oidcReq.auth != nil {
 		return logical.ErrorResponse(errLoginFailed + " Expired or missing OAuth state."), nil
 	}
 
-	clientNonce := d.Get("client_nonce").(string)
-
-	// If a client_nonce was provided at the start of the auth process as part of the auth_url
-	// request, require that it is present and matching during the callback phase.
-	if oidcReq.clientNonce != "" && clientNonce != oidcReq.clientNonce {
-		return logical.ErrorResponse("invalid client_nonce"), nil
-	}
-
 	roleName := oidcReq.rolename
 	role, err := b.role(ctx, req.Storage, roleName)
 	if err != nil {
+		b.deleteOIDCRequest(stateID)
 		return nil, err
 	}
 	if role == nil {
+		b.deleteOIDCRequest(stateID)
 		return logical.ErrorResponse(errLoginFailed + " Role could not be found"), nil
 	}
 
+	useHttp := false
```

Comment:
Do we just want to simplify this to `useHttp := role.CallbackMode == callbackModeDirect` and rename it appropriately (`isDirectCallback`)? I think the `useHttp`'s use further down as a sentinel for this makes it a touch confusing, and parameter name on `loginFailedResponse` is clear enough that that's what the first argument means, without naming every use of this `useHttp`.

Reply:
No, because what is mostly for is to indicate whether or not to use http in a `loginFailedResponse()` call.

I replaced the two cases where it was used for other purposes to instead rely on a check for `role.CallbackMode == callbackModeDirect`.